runtime/gc_modron_startup/mminit.cpp

```diff
@@ -3054,7 +3054,21 @@ gcReinitializeDefaultsForRestore(J9VMThread* vmThread)
 	bool result = true;
 
 	PORT_ACCESS_FROM_JAVAVM(vm);
-
+	OMRPORT_ACCESS_FROM_J9PORT(PORTLIB);
+	/* Note here we update this parameter which represents the machine physical memory,
+	 * but not the original heap geometry from the snapshot run. We just force the heap
+	 * not to go beyond a certain thereshold calculated based on the new (restore) memory
+	 * availalibity through setting SoftMx.
+	 */
+	extensions->usablePhysicalMemory = omrsysinfo_get_addressable_physical_memory();
+	/* We use false here for computeDefaultMaxHeapForJava(), since the restore
+	 * path is only active for releases after Java 8 releases
+	 */
+	uintptr_t candidateSoftMx = extensions->computeDefaultMaxHeapForJava(false);
+	/* if snapshot machine's memory is higher than the restore machine */
+	if (extensions->memoryMax > candidateSoftMx) {
```

Comment:
theoretically, a  more optimal way to deal with this was to first check if physical memory was reduced, end if so invoke computeDefaultMaxHeapForJava and store it directly to extensions->softMx. the difference is that we would not invoke computeDefaultMaxHeapForJava not needed. in practice, probably it's a negligible difference

Reply:
@dmitripivkine any preference?
also, please re-review the whole change

Reply:
I have no preference. However what about the case when `-Xsoftmx` is set in the command line? I think it should be handled too.

Reply:
Except this change looks correct. I can do final review when ready

Reply:
you mean if it was provided on snapshot side already?
not likely scenario, but if it is the case, I guess we should not be overwritting it with a higher value
so if 0 != ext->softMx and candidateSoftMx > ext->softMx, don't set it

Reply:
Is it possible that the computeDefaultMaxHeapForJava() takes different path to utilize physical memory inside here?

Reply:
I think we can keep the current setup:

Let's say we snapshot on a 8GB machine (use 3/4 as the coefficient inside computeDefaultMaxHeapForJava() from some decision path) and restore on a 16GB machine (use 1/4 as the coefficient inside computeDefaultMaxHeapForJava() from some other decision paths). If we just compare the physical memory, then 8GB < 16GB so no need to set softMx. But in terms of the actual value, the memoryMax from snapshot is 8GB * 3/4 = 6GB, while the candidateSoftMx = 16GB * 1/4 = 4GB < 6GB, and we should set softMx.

Reply:
Also add logic for checking the previously set softMx

Reply:
we should also add a comparison with minimum/initial heap size (as set by snapshot run via Xms option). we should refuse to run if softMx ends up being lower than it
similar to just a plain run would refuse:

```
./java -Xmx2G -Xms256M -Xsoftmx128M -verbose:gc,sizes

JVMJ9GC017E -Xsoftmx too small, must be at least 256 Mbytes
JVMJ9VM015W Initialization error for library j9gc29(2): Failed to initialize
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```